### PR TITLE
Introduce support for authenticated docker registries using imagePullSecrets, Fixes #4413

### LINKED ISSUE #4413: Adding support for Docker Private Virtual Registries like Jfrog Artifactory for Airgap systems.
### What problem are you trying to solve?
Jfrog Artifactory or other private artifact registries for docker container are common in airgap systems, where the Linkerd chart referencing docker images, cannot be downloaded from the default upstream urls ( example: gcr.io/linkerd-io/proxy ).

Also these private registries are configured with authentication, for which the current implementation has no support.

### How should the problem be solved?
By adding `imagePullSecrets: <registry-secret>` parameter in the values.yaml (which is not implemented and therefore needs code change, the airgap systems can smoothly install the chart. 

The implementation should assume the registry secret is already available with the namespace.

User has to configure the value for `imagePullSecrets` in Linkerd Chart's values.yaml, and change the image reference of docker images in the values.yaml to point to their Jfrog Registry so that the charts can directly be deployed. 

### Any alternatives you've considered?
The other way to support such airgap system is to download the chart in zip format, use helm package command to package it after code changes as shared above, and deploy it in Jfrog Artifactory as part of local chart repository. This is a tedious process.

### How would users interact with this feature?
User will supply the `ImagePullSecrets` parameter to configure their artifactory registry, along with changing the reference of docker images in the values.yaml

https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account

### PR DESCRIPTION
**Pull request by c-n-c**
Introduce support for authenticated docker registries using imagePullSecrets 

Problem: Private Docker Registries are not supported for the moment as detailed in issue #4413

Solution: Every Service Account of linkerd subcomponents are Attached with imagePullSecrets,
which in turn can then pulls the docker images from authenticated private registries using them.
The imagePullSecret is configured in global.imagePullSecret parameter of values.yaml like

imagePullSecret:
  - name: <name-of-private-registry-secret-resource>

Fixes #4413

Signed-off-by: Nilakhya <nilakhya@hotmail.com>

### CODE REVIEW

File: charts/add-ons/grafana/templates/grafana-rbac.yaml [hunk 1]
```diff
@@ -11,3 +11,4 @@ metadata:
   labels:
     {{.Values.global.controllerComponentLabel}}: grafana
     {{.Values.global.controllerNamespaceLabel}}: {{.Values.global.namespace}}
+{{ include "partials.image-pull-secrets" . }}
```
Comment (alpeb) [inline charts/add-ons/grafana/templates/grafana-rbac.yaml:14]: Please add use `{{- include "partials.image-pull-secrets" . }}` to suppress line-feeds/spaces (this is what's making the tests fail)

File: charts/linkerd2-cni/templates/cni-plugin.yaml [hunk 2]
```diff
@@ -56,6 +56,7 @@ metadata:
   namespace: {{.Values.namespace}}
   labels:
     {{.Values.cniResourceLabel}}: "true"
+{{ include "partials.image-pull-secrets" . }}
```
Comment (alpeb) [inline charts/linkerd2-cni/templates/cni-plugin.yaml:59]: `charts/linkerd2-cni` is a separate chart that doesn't receive the global config from `charts/linkerd2/values.yml`. In order to use `global.imagePullSecrets` in this chart, you need to declare it in `charts/linkerd2-cni/values.yaml`.

File: charts/linkerd2/values.yaml [hunk 3]
```diff
@@ -92,6 +92,13 @@ global:
   linkerdNamespaceLabel: linkerd.io/is-control-plane
   workloadNamespaceLabel: linkerd.io/workload-ns
 
+  #
+  ## Define your image pull secrets for private docker registries like the following
+  #imagePullSecrets:
```
Comment (alpeb) [inline charts/linkerd2/values.yaml:97]: In this comment, please clarify that this is for adding the ImagePullSecret to each of the control plane service accounts (and not adding the ImagePullSecret to each Pod).

File: charts/add-ons/grafana/templates/grafana-rbac.yaml [hunk 4]
```diff
@@ -11,3 +11,7 @@ metadata:
   labels:
     {{.Values.global.controllerComponentLabel}}: grafana
     {{.Values.global.controllerNamespaceLabel}}: {{.Values.global.namespace}}
+{{- if .Values.global.imagePullSecrets }}
```
Comment (Pothulapati) [inline charts/add-ons/grafana/templates/grafana-rbac.yaml:14]: because `partials` is also a sub-chart for add-ons, `{{- include "partials.image-pull-secrets" . }}` should work directly

**Review by Pothulapati — APPROVED** [review]
I tested this out by pushing images into the new GitHub package registry.

If you are trying to do the same, Make sure to push and pull the images into `ghcr.io` instead of `docker.pkg.github.com` as the later act weird as per https://github.com/containerd/containerd/issues/3291#issuecomment-685099010

File: charts/add-ons/grafana/templates/grafana-rbac.yaml (hunk 4 above)
Comment (c-n-c) [reply charts/add-ons/grafana/templates/grafana-rbac.yaml:14]: I tried this, but were failing tests so on a later commit i  introduced this.

File: charts/linkerd2-cni/values.yaml [hunk 5]
```diff
@@ -19,3 +19,15 @@ priorityClassName: ""
...
+
+#
+## For Private docker registries, authentication is needed.
+# If the control plane service images are pulled from a
+# protected docker registry, define pull secrets as follows:
+#
+#imagePullSecrets:
+#    - name: my-private-docker-registry-login-secret
+#
+# The pull secrets are applied to the respective service accounts
```
[the remark this answers is not in the document]
Comment (c-n-c) [reply charts/linkerd2-cni/values.yaml:31]: I just revamped the comments as per @alpeb in his last review.

File: charts/partials/templates/_pull-secrets.tpl [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{{- define "partials.image-pull-secrets" -}}
+{{- if .Values.global.imagePullSecrets }}
```
[the remark this answers is not in the document]
Comment (c-n-c) [reply charts/partials/templates/_pull-secrets.tpl:2]: maintaining homogenousity

File: charts/linkerd2-cni/templates/cni-plugin.yaml [hunk 7]
```diff
@@ -57,6 +57,10 @@ metadata:
   namespace: {{.Values.namespace}}
   labels:
     {{.Values.cniResourceLabel}}: "true"
+{{- if .Values.imagePullSecrets }}
```
[the remark this answers is not in the document]
Comment (c-n-c) [reply charts/linkerd2-cni/templates/cni-plugin.yaml:60]: As per @alpeb in last review,  cni plugin does not import partials, so i included it

File: charts/add-ons/prometheus/templates/prometheus-rbac.yaml [hunk 8]
```diff
@@ -39,3 +39,7 @@ metadata:
   labels:
     {{.Values.global.controllerComponentLabel}}: prometheus
     {{.Values.global.controllerNamespaceLabel}}: {{.Values.global.namespace}}
+{{- if .Values.global.imagePullSecrets }}
```
[the remark this answers is not in the document]
Comment (c-n-c) [reply charts/add-ons/prometheus/templates/prometheus-rbac.yaml:42]: I had to as i shared above, the tests start failing otherwise

File: charts/linkerd2-cni/values.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (alpeb) [reply charts/linkerd2-cni/values.yaml:31]: I agree with @Pothulapati that the description is too verbose. My suggestion was just that you clarified that these are ServiceAccount ImagePullSecrets, and not Pod ImagePullSecrets. Also you can provide a commented example like so:

```
  imagePullSecrets: []
  # - name: my-private-docker-registry-login-secret                                                                 
```

File: charts/add-ons/tracing/values.yaml [hunk 9]
```diff
@@ -1 +1 @@
-# Default values for tracing.
\ No newline at end of file
+# Default values for tracing.
```
Comment (alpeb) [inline charts/add-ons/tracing/values.yaml:1]: Would you please revert this change?

File: charts/add-ons/prometheus/values.yaml [hunk 10]
```diff
@@ -13,3 +13,4 @@ globalConfig:
   scrape_interval: 10s
   scrape_timeout: 10s
   evaluation_interval: 10s
+
```
Comment (alpeb) [inline charts/add-ons/prometheus/values.yaml:16]: Same here

File: charts/add-ons/grafana/templates/grafana-rbac.yaml (hunk 4 above)
Comment (alpeb) [reply charts/add-ons/grafana/templates/grafana-rbac.yaml:14]: Just using `{{- include "partials.image-pull-secrets" . }}` works for me... Can you push it just with that, and so we can see what's the actual failure in CI?

**Review by alpeb — COMMENTED** [review]
In order to have this work from the install CLI and not just Helm, you need to add this to `./pkg/charts/linkerd2/values.go` under the `Global` struct:
```
ImagePullSecrets         []map[string]string `json:"imagePullSecrets""`
```

File: charts/add-ons/prometheus/values.yaml (hunk 10 above)
Comment (c-n-c) [reply charts/add-ons/prometheus/values.yaml]: @apleb done

File: charts/add-ons/tracing/values.yaml (hunk 9 above)
Comment (c-n-c) [reply charts/add-ons/tracing/values.yaml]: @apleb done

File: charts/add-ons/grafana/templates/grafana-rbac.yaml (hunk 4 above)
Comment (c-n-c) [reply charts/add-ons/grafana/templates/grafana-rbac.yaml]: @apleb, @Pothulapati  done

**Comment by Pothulapati** [comment]
@c-n-c I think we are in the right direction with defining `image-pull-secrets` tpl again for `cni`, as its config structure is different i.e it does not have `global` while others do. So, It makes sense to have a separate tpl for CNI. I also see that the PR is failing because of `gofmt`, Can you please run gofmt on top of the project so that it follows the required structure?

@alpeb Hmm. even though we are adding that property to `Values` it's still not usable through the CLI, right? :thinking: unless we add a relevant CLI `installOption` field.

**Comment by alpeb** [comment]
> @alpeb Hmm. even though we are adding that property to `Values` it's still not usable through the CLI, right?  unless we add a relevant CLI `installOption` field.

Correct not usable in the CLI, but we still want the CLI to pick the default values set in `values.yaml`. So if in the future we decide to change any default in `values.yaml`, the CLI will pick the change as well, even if there's no flag for it. This can be tested by changing the default, recompiling the CLI, and running `linkerd install` again to see the change was picked up.

**Comment by alpeb** [comment]
@c-n-c please reset `actual.Global.ImagePullSecrets = nil` in `TestNewValues` for the unit test to pass :wink:

**Review by alpeb — APPROVED** [review]
Thanks @c-n-c , this looks good to me.

**Comment by alpeb** [comment]
@zaharidichev do you mind giving a look at this one?

**Review by zaharidichev — APPROVED** [review]
LGTM

**Comment by alpeb** [comment]
Thanks @c-n-c :+1: 
I took the liberty of fixing the merge conflicts before merging to main.
